horovod/spark/common/util.py

```diff
@@ -0,0 +1,545 @@
+                continue
+
+            input_size = abs(np.prod(input_shape))
+            if col_size != input_size:
+                raise ValueError(
+                    'Feature column \'{col}\' with size {feature} must equal that of the '
+                    'model input at index {idx} with size {input}'
+                    .format(col=col, feature=col_size, idx=idx, input=input_size))
+
+    if output_shapes is not None:
+        label_count = len(label_columns)
+        if label_count != len(output_shapes):
+            raise ValueError('Label column count {labels} must equal '
+                             'model outputs count {outputs}'
+                             .format(labels=label_count, outputs=len(output_shapes)))
+
+        for idx, col, output_shape in zip(range(label_count), label_columns, output_shapes):
+            col_size = metadata[col]['shape']
+            output_size = abs(np.prod(output_shape))
+            if col_size != output_size:
+                raise ValueError('Label column \'{col}\' with size {label} must equal that of the '
+                                 'model output at index {idx} with size {output}'
+                                 .format(col=col, label=col_size, idx=idx, output=output_size))
+
+
+def _get_col_info(df):
+    """
+    Infer the type and shape of all the columns.
+
+    NOTE: This function processes the entire DataFrame, and can therefore be very expensive to run.
+
+    TODO(travis): Oonly run this if user sets compress_sparse param, otherwise convert all to Array.
+    """
+
+    def get_meta(row):
+        row_dict = row.asDict()
+        row_schema = []
+        for col_name, data_col in row_dict.items():
+            dtype = type(data_col)
+            if isinstance(data_col, DenseVector):
+                # shape and size of dense vector are the same
+                shape = size = data_col.array.shape[0]
+            elif isinstance(data_col, SparseVector):
+                # shape is the total size of vector
+                shape = data_col.size
+                # size is the number of nonzero elements in the sparse vector
+                size = data_col.indices.shape[0]
+            elif isinstance(data_col, list):
+                shape = size = len(data_col)
+            elif isinstance(data_col, type(None)):
+                # Python 2.7 compat: NoneType is not pickleable
+                # see: https://bugs.python.org/issue6477
+                dtype = NullType
+                shape = size = 1
+            else:
+                shape = size = 1
+            row_schema.append((col_name, ({dtype}, {shape}, {size})))
+        return row_schema
+
+    def merge(x, y):
+        dtypes, shapes, sizes = x
+        new_dtypes, new_shapes, new_sizes = y
+        dtypes.update(new_dtypes)
+        shapes.update(new_shapes)
+        sizes.update(new_sizes)
+        return dtypes, shapes, sizes
+
+    raw_col_info_list = df.rdd.flatMap(get_meta).reduceByKey(merge).collect()
```

Comment:
I think we should not collect this data directly. It should be a big data if the data format is `SparseVector`. The following check and merge process should be done in spark. We just need to collect the data `pair (shape, max_size)`.

Reply:
Good point. I changed the `merge` function to only track the min and max sizes and shapes, which will give us all the same information (max size, are the shapes and sizes all uniform), but requires only tracking at most two elements instead of unbounded.